### ISSUE TITLE
lodash-es - type: module in package.json breaks npm install with some NPM versions

### ISSUE THREAD

**Issue by phistuck** [lodash/lodash#4359]
We get
```
09:45:34 npm ERR! code EMISSINGARG
09:45:34 
09:45:34 npm ERR! typeerror Error: Missing required argument #1
09:45:34 npm ERR! typeerror     at andLogAndFinish (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:31:3)
09:45:34 npm ERR! typeerror     at fetchPackageMetadata (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:51:22)
09:45:34 npm ERR! typeerror     at resolveWithNewModule (/usr/local/lib/node_modules/npm/lib/install/deps.js:490:12)
09:45:34 npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/install/deps.js:491:7
09:45:34 npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/iferr/index.js:13:50
09:45:34 npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:37:12
09:45:34 npm ERR! typeerror     at addRequestedAndFinish (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:82:5)
09:45:34 npm ERR! typeerror     at returnAndAddMetadata (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:118:7)
09:45:34 npm ERR! typeerror     at pickVersionFromRegistryDocument (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:135:20)
09:45:34 npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/iferr/index.js:13:50
09:45:34 npm ERR! typeerror This is an error with npm itself. Please report this error at:
09:45:34 npm ERR! typeerror     <http://github.com/npm/npm/issues>
```
When running `npm i lodash-es@4.17.13` using certain Node/NPM versions.

Please, remove `"type": "module"` from package.json. This is affecting over 500 modules.

**Comment by phistuck** [comment]
This is a follow up to the closed https://github.com/lodash/lodash/issues/4358

**Comment by phistuck** [comment]
@falsyvalues - maybe revert this in a new version (4.17.14) and change the version to 5.0.0 for this change because it breaks random NPM versions?
(Sorry for the duplicate, not sure which issue is more appropriate now)

**Comment by Unsfer** [comment]
Same error for me on node 4.6.1 and npm 3.8.7. Production install broken :-(
Works on npm 3.10.7

**Comment by whateweryouwant** [comment]
Same here, production bulid broken.
[11:30:18] [Step 1/1] [ERROR] npm ERR! node v6.7.0
[11:30:18] [Step 1/1] [ERROR] npm ERR! npm  v3.10.3
I prefer what @phistuck proposed.

**Comment by falsyvalues** [comment]
I don't think that relaying on unsupported `npm` version is a way to go and valid reason for rolling back this. But its up to @jdalton.

**Comment by phistuck** [comment]
@falsyvalues and @jdalton -
Looks like it also happens with Node 10 and NPM 6.9 which should be supported -
https://github.com/lodash/lodash/issues/4358#issuecomment-509963124

So it would be great if you reconsidered, too.

**Comment by Unsfer** [comment]
Yeah, let's publish minor version of mass used package with breaking changes and say that folks need to update their environment

**Comment by falsyvalues** [comment]
It's not breaking change because it does literally nothing to lodash itself. `npm` is the thing to blame here.

**Comment by Unsfer** [comment]
I think we're all here not for searching for who to blame but to find less painful solution.
So I to prefer what @phistuck proposed if it can be done.

**Comment by jdalton** [comment]
Related to  PR #4292 and issue #4358. Thanks for the feedback y'all. I'm not only the creator of Lodash, but I'm also a Node core member and a member of the Node Module Working Group. I'll give the working group feedback that the `type` field breaks folks at the moment. 

Could one of you all follow-up with an `npm` issue for `type` that would rock 🤘!
I'll revert the `type` addition for now.

**Comment by falsyvalues** [comment]
> @falsyvalues and @jdalton -
> Looks like it also happens with Node 10 and NPM 6.9 which should be supported -
> [#4358 (comment)](https://github.com/lodash/lodash/issues/4358#issuecomment-509963124)
> 
> So it would be great if you reconsidered, too.

If so, for sure. I could not repro that on node v10.14.1 and npm v6.9.0 but will look more.

**Comment by phistuck** [comment]
@jdalton - thank you very much!

> Could one of you all follow-up with an `npm` issue for `type` that would rock 🤘!
@horaceleung - if the fix resolves your case, it would be great if you followed up as suggested because I am not using such a new version of NPM/Node.js (unfortunately), so I cannot file from my experience.

**Comment by jdalton** [comment]
`lodash-es@4.17.14` is published :tada:

**Comment by falsyvalues** [comment]
@jdalton Thanks for stepping in 🥇

**Comment by whateweryouwant** [comment]
speed of light! great.

**Comment by horaceleung** [comment]
Thanks very much!! :)

**Comment by GeoffreyBooth** [comment]
Here’s a one-liner to reproduce the issue:

```
✦ docker run node:6.7.0 bash -c 'npm install lodash-es@4.17.13'
npm info it worked if it ends with ok
npm info using npm@3.10.3
npm info using node@v6.7.0
npm info attempt registry request try #1 at 11:44:33 PM
npm http request GET https://registry.npmjs.org/lodash-es
npm http 200 https://registry.npmjs.org/lodash-es
npm ERR! Linux 4.9.125-linuxkit
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "i" "lodash-es@4.17.13"
npm ERR! node v6.7.0
npm ERR! npm  v3.10.3
npm ERR! code EMISSINGARG

npm ERR! typeerror Error: Missing required argument #1
```

However the problem goes away in the latest version of Node 6 (which reached end-of-life back in April, and is currently unsupported) which shipped with npm 3.10.10:

```
✦ docker run node:6.17.1 bash -c 'npm init --yes > /dev/null; npm i lodash-es@4.17.13'
@1.0.0 /
`-- lodash-es@4.17.13

npm WARN @1.0.0 No description
npm WARN @1.0.0 No repository field.
```

And the problem is also not present in the oldest currently supported version of Node, 8.0.0 (with npm 5.0.0):

```
✦ docker run node:8.0.0 bash -c 'npm install lodash-es@4.17.13'
npm info it worked if it ends with ok
npm info using npm@5.0.0
npm info using node@v8.0.0
npm http fetch GET 200 https://registry.npmjs.org/lodash-es 6976ms
npm http fetch GET 200 https://registry.npmjs.org/lodash-es/-/lodash-es-4.17.13.tgz 4046ms
npm info lifecycle lodash-es@4.17.13~preinstall: lodash-es@4.17.13
npm info linkStuff lodash-es@4.17.13
npm info lifecycle lodash-es@4.17.13~install: lodash-es@4.17.13
npm info lifecycle lodash-es@4.17.13~postinstall: lodash-es@4.17.13
...
added 1 package in 11.482s
npm info ok
```

Someone could write a script based on the above to test every npm version to see which ones have this issue, but based on this at least it’s fair to assume npm < 3.10.10 might be affected. Here are some workarounds a Node 6.7.0 user could do in order to install `lodash-es@4.17.13`:

- Upgrade npm, if possible. Especially to a version that supports `package-lock.json`, so that exact versions of dependencies (and their dependencies) are installed, rather than the latest of a particular semver major/minor line. `package-lock.json` was introduced in npm 5.0.0.

- Use yarn. `docker run node:6.7.0 bash -c 'npm i -g yarn; yarn add lodash-es@4.17.13'` does not have this issue.

- Commit your `node_modules` folder into your repo, or archive it somewhere and have it copied into place by your build script. Presumably a Node 6 app or package isn’t still under active development, and shouldn’t be receiving the latest versions of all dependencies. You probably should lock down all versions of all dependencies, at least until the next time you’re working on the app (and can use a newer version of npm).

- Set up a mirror of the npm registry using a tool like [Verdaccio](https://verdaccio.org/) and control the package versions you want hosted there, and have your build script npm use that registry instead.

Node 6 was end-of-lifed earlier than usual for security reasons, so it’s really not a good idea to still be running it in production; and there’s no reason that I can think of to still be running npm 3.10.3 other than the understandable pain of upgrading. At least upgrading npm or switching to yarn shouldn’t require any changes to app code, unlike (potentially) upgrading Node.

**Comment by speller** [comment]
Lodash-es can not be used with Node + ES Modules without "type: module". That's funny - ES version of lodash doesn't work with ES modules :)
